File: gtci/04-merge-intervals/python3/6-employee-free-time.py
import heapq


class Interval:
    def __init__(self, start, end):
        self.start = start
        self.end = end


class EmployeeInterval:

    def __init__(self, interval, employeeIndex, intervalIndex):
        self.interval = interval  # interval representing employee's working hours
        # index of the list containing working hours of this employee
        self.employeeIndex = employeeIndex
        self.intervalIndex = intervalIndex  # index of the interval in the employee list

    def __lt__(self, other):
        # min heap based on meeting.end
        return self.interval.start < other.interval.start

    def __str__(self):
        return f"({self.employeeIndex}: ({self.intervalIndex}, {self.interval}))"

    def __repr__(self):
        return self.__str__()


class Solution:
    def findEmployeeFreeTime(self, schedule):
        result = []

        min_heap = []

        for idx, employee_schedule in enumerate(schedule):
            heapq.heappush(min_heap, EmployeeInterval(employee_schedule[0], idx, 0))

        previous_interval = min_heap[0].interval

        while min_heap:
            top = heapq.heappop(min_heap)
            earliest_start_interval = top.interval

            if previous_interval.end < earliest_start_interval.start:
                result.append([previous_interval.end, earliest_start_interval.start])
                previous_interval = earliest_start_interval
            else:
                previous_interval.start = min(
                    previous_interval.start, earliest_start_interval.start
                )
                previous_interval.end = max(
                    previous_interval.end, earliest_start_interval.end
                )

            # add more of the employee's intervals if they have any
            if top.intervalIndex + 1 < len(schedule[top.employeeIndex]):
                heapq.heappush(
                    min_heap,
                    EmployeeInterval(
                        schedule[top.employeeIndex][top.intervalIndex + 1],
                        top.employeeIndex,
                        top.intervalIndex + 1,
                    ),
                )

        return result


if __name__ == "__main__":
    sol = Solution()

    assert sol.findEmployeeFreeTime(
        [[Interval(1, 3), Interval(5, 6)], [Interval(2, 3), Interval(6, 8)]]
    ) == [[3, 5]]

    assert sol.findEmployeeFreeTime(
        [[Interval(1, 3), Interval(9, 12)], [Interval(2, 4), Interval(6, 8)]]
    ) == [[4, 6], [8, 9]]

    assert sol.findEmployeeFreeTime(
        [[Interval(1, 3), Interval(2, 4)], [Interval(3, 5), Interval(7, 9)]]
    ) == [[5, 7]]

    print("All test cases passed.")

File: gtci/04-merge-intervals/cpp/6-employee-free-time.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <queue>
#include <vector>

#include "interval.h"

using namespace std;

class Solution {
   public:
    vector<Interval> findEmployeeFreeTime(
        const vector<vector<Interval>> &schedule) {
        vector<Interval> result;

        auto employeeCompare = [](const pair<Interval, pair<int, int>> &x,
                                  const pair<Interval, pair<int, int>> &y) {
            return x.first.start > y.first.start;
        };

        std::priority_queue<std::pair<Interval, std::pair<int, int>>,
                            vector<std::pair<Interval, std::pair<int, int>>>,
                            decltype(employeeCompare)>
            minHeap(employeeCompare);

        for (size_t i = 0; i < schedule.size(); i++) {
            const vector<Interval> &curr = schedule[i];
            minHeap.push({curr[0], {i, 0}});
        }

        Interval previousInterval = minHeap.top().first;
        while (!minHeap.empty()) {
            auto top = minHeap.top();
            minHeap.pop();

            if (previousInterval.end < top.first.start) {
                result.push_back({previousInterval.end, top.first.start});
                previousInterval = top.first;
            } else {
                previousInterval.start =
                    std::min(previousInterval.start, top.first.start);
                previousInterval.end =
                    std::max(previousInterval.end, top.first.end);
            }

            // add more of the employee's intervals if they have any
            if (top.second.second + 1 < schedule[top.second.first].size()) {
                const vector<Interval> &intervals = schedule[top.second.first];
                minHeap.push(
                    {intervals[top.second.second + 1],
                     std::pair(top.second.first, top.second.second + 1)});
            }
        }

        return result;
    }
};

int main() {
    Solution sol;

    auto result1 = sol.findEmployeeFreeTime(
        {{Interval(1, 3), Interval(5, 6)}, {Interval(2, 3), Interval(6, 8)}});
    assert(result1 == vector<Interval>{Interval(3, 5)});

    auto result2 = sol.findEmployeeFreeTime(
        {{Interval(1, 3), Interval(9, 12)}, {Interval(2, 4), Interval(6, 8)}});
    assert(result2 == vector<Interval>({Interval(4, 6), Interval(8, 9)}));

    auto result3 = sol.findEmployeeFreeTime(
        {{Interval(1, 3), Interval(2, 4)}, {Interval(3, 5), Interval(7, 9)}});
    assert(result3 == vector<Interval>{Interval(5, 7)});

    cout << "All test cases passed." << endl;

    return 0;
}
